File: 46-permutations/46-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        if nums is None or len(nums) == 0:
            return []
        if len(nums) == 1:
            return [[nums[0]]]
        if len(nums) == 2:
            return [[nums[0],nums[1]], [nums[1],nums[0]]]
        result = []
        for num in nums:
            tmp = nums.copy()
            tmp.remove(num)
            pre_res = self.permute(tmp)
            for arr in pre_res:
                result.append([num] + arr)
            
        return result

File: 46-permutations/46-permutations.kt
class Solution {
 fun permute(nums: IntArray): List<List<Int>> {
        var ans = ArrayList<ArrayList<Int>>()
        
        fun track(level: ArrayList<Int>){
            if(level.size == nums.size){
                var temp = level.toSet()
                if(temp.size == nums.size)
                    ans.add(ArrayList(level))
                    
                return 
            }
            
            for(i in nums){
                level.add(i)
                track(level)
                level.removeAt(level.size-1)
            }
        }
    
        track(ArrayList<Int>())
        
        return ans.toList()
    }
}